Client/Application/AppState.swift

```diff
@@ -0,0 +1,29 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+protocol AppStateDelegate: class {
+    func appDidUpdateState(appState: AppState)
+}
+
+enum AppState {
+    case Tab(tabState: TabState)
```

Comment:
How would you have it instead. It's either an enum, or it's a struct. It's an enum as that was how we discussed with @rnewman it should be  - totally immutable describing a snapshot in time of state. We could make this a struct instead, containing a location (tab, home panels, tab tray) like we had before, but my memory of that conversation was that it shouldn't be that way.

Also, if we don't remember the state and use it within the menu to determine the configuration/display options, how should we use it? Split out the information within the app state into vars within the menu and then query those? Have the MenuConfiguration remember the app state and just return the right info (so instead of calling iconForState() you just call icon() and the configuration knows what to return because it understands the app state?) Some other way?